### PR TITLE
data: improve element access to have constant amortized cost

### PR DESCRIPTION
**Pull request by daejunpark**
Now `#drop` is defined in a way that utilizes the function evaluation cache, thus the amortized  complexity of `PGM [ PCOUNT ]` becomes constant.

The disadvantage of this new `#drop` is that it is not tail-recursive. But its overhead is relatively small and acceptable for the symbolic execution.  For the concrete execution, `#drop` is only used for `LOG(N)` where `N <= 4`, so it should be fine.

### CODE REVIEW

File: data.md [hunk 1]
```diff
@@ -420,8 +420,9 @@ A cons-list is used for the EVM wordstack.
     syntax WordStack ::= #drop ( Int , WordStack ) [function, functional]
  // ---------------------------------------------------------------------
     rule #drop(N, WS)         => WS                  requires notBool N >Int 0
-    rule #drop(N, .WordStack) => .WordStack
-    rule #drop(N, (W : WS))   => #drop(N -Int 1, WS) requires N >Int 0
+    rule #drop(N, WS)         => #drop(1, #drop(N -Int 1, WS)) requires N >Int 1
+    rule #drop(1, _ : WS)     => WS
+    rule #drop(1, .WordStack) => .WordStack
```
Comment (ehildenb) [inline data.md]: This is no longer a total function, so cannot have `functional`, or you need to make it a total function.

Comment (daejunpark) [reply data.md]: Um, I think it is still total.  But in any case, this has another issue, so I'll update it later.

Comment (daejunpark) [reply data.md]: BTW, how to make this PR "draft"?

Comment (daejunpark) [reply data.md]: Never mind. Now it is ready.

**Comment by daejunpark** [comment]
@ehildenb This is ready to review.  Note that we can keep only the symbolic version, and drop the concrete version, if you prefer.

**Comment by ehildenb** [comment]
```

```* a3582de2 - data: drop concrete version of #drop                                             - Daejun Park (HEAD, upstream/fix-symbolic-bytes-access, fix-symbolic-bytes-access)
| testset: perf - Mon 25 Nov 2019 12:14:14 PM MST
| 0  0.37s     0.00s      0.35s     3420KB      rm    -rf               deps                  .build/defn
| 0  1.27s     0.81s      0.51s     28184KB     git   submodule         update                --init                      --recursive
| 0  239.28s   1084.71s   1088.58s  2571392KB   make  deps              K_BUILD_TYPE=Release
| 0  100.62s   650.22s    15.45s    2404712KB   make  build-java        build-haskell         build-llvm                  -j8          LIBFF_CC=clang  LIBFF_CXX=clang++
| 0  139.92s   1251.84s   338.19s   18960440KB  make  test-conformance  -j12                  TEST_CONCRETE_BACKEND=llvm
| 0  1466.94s  21716.05s  525.63s   3400456KB   make  test-prove        -j10
| 
* 9f2512eb - data: improve element access to have constant amortized cost                     - Daejun Park
| testset: perf - Mon 25 Nov 2019 11:41:45 AM MST
| 0  0.37s     0.01s      0.34s     3468KB      rm    -rf               deps                  .build/defn
| 0  1.29s     0.85s      0.48s     28092KB     git   submodule         update                --init                      --recursive
| 0  239.34s   1090.09s   1093.18s  2548580KB   make  deps              K_BUILD_TYPE=Release
| 0  101.93s   660.59s    14.85s    2403788KB   make  build-java        build-haskell         build-llvm                  -j8          LIBFF_CC=clang  LIBFF_CXX=clang++
| 0  139.42s   1246.14s   338.02s   18960344KB  make  test-conformance  -j12                  TEST_CONCRETE_BACKEND=llvm
| 0  1486.78s  21426.74s  509.09s   3719308KB   make  test-prove        -j10
| 
* cd400802 - Factor out JSON sanitization steps (#569)                                        - ehildenb
  testset: perf - Mon 25 Nov 2019 11:08:55 AM MST
  0  0.34s     0.00s      0.32s     3364KB      rm    -rf               deps                  .build/defn
  0  1.27s     0.87s      0.44s     27956KB     git   submodule         update                --init                      --recursive
  0  237.54s   1056.54s   1104.28s  2561908KB   make  deps              K_BUILD_TYPE=Release
  0  99.82s    641.95s    14.30s    2274656KB   make  build-java        build-haskell         build-llvm                  -j8          LIBFF_CC=clang  LIBFF_CXX=clang++
  0  139.13s   1241.40s   336.55s   18960404KB  make  test-conformance  -j12                  TEST_CONCRETE_BACKEND=llvm
  0  1503.32s  23267.59s  547.91s   4485100KB   make  test-prove        -j10

**Comment by ehildenb** [comment]
We see about an 8% performance improvement for proving, and about a 20-25% memory usage improvement on proving!

No real difference for concrete execution.

**Pull request merged**